### PR TITLE
Style/EmptyLiteral autocorrect bugfix

### PR DESCRIPTION
**Pull request by ota42y**
Style/EmptyLiteral autocorrect `Array.new` to `[]`.

This code is valid code
`Array.new { 1 }`
but, autocorrect change to this and it's invalid code.
`[] { 1 }`

So, I change ignore autocorrect.
I think, this is bad code (`Array.new { 1 }` equal to `[]`) so I fix autocorrect only.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/empty_literal.rb [hunk 1]
```diff
@@ -36,7 +38,8 @@ def on_send(node)
 
         def autocorrect(node)
           lambda do |corrector|
-            corrector.replace(replacement_range(node), correction(node))
+            result = correction(node)
+            corrector.replace(replacement_range(node), result) if result
```
Comment (pocke) [inline lib/rubocop/cop/style/empty_literal.rb]: The check for `result` should be outside the `lambda`.
By `autocorrect` method returns a nil, RuboCop treats it as not being auto-corrected.

For example: https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/lint/safe_navigation_chain.rb#L48-L56

**Comment by ota42y** [comment]
fixed!

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,7 @@
 * [#4019](https://github.com/bbatsov/rubocop/pull/4019): Make configurable `Style/MultilineMemoization` cop. ([@pocke][])
 * [#4018](https://github.com/bbatsov/rubocop/pull/4018): Add autocorrect `Lint/EmptyEnsure` cop. ([@pocke][])
 * [#4028](https://github.com/bbatsov/rubocop/pull/4028): Add new `Style/IndentHeredoc` cop. ([@pocke][])
+* [#4049](https://github.com/bbatsov/rubocop/pull/4049): Bugfix for `Style/EmptyLiteral` cop. ([@ota42y][])
```
Comment (bbatsov) [inline CHANGELOG.md]: This should be in the "bugs fixed" section and be a bit more descriptive.

File: spec/rubocop/cop/style/empty_literal_spec.rb [hunk 3]
```diff
@@ -29,6 +29,24 @@
       new_source = autocorrect_source(cop, 'test = Array.new')
       expect(new_source).to eq('test = []')
     end
+
+    it 'not auto-corrects Array.new with block' do
```
Comment (bbatsov) [inline spec/rubocop/cop/style/empty_literal_spec.rb]: not auto-corrects -> does not auto-correct

**Comment by bbatsov** [comment]
Shouldn't you actually fix the offense detection logic instead? If this node wasn't marked as an offense it wouldn't be auto-corrected at all.

**Comment by ota42y** [comment]
I'm fixed!

I think this should be offense.
Because
 - `Array.new { 1 }` is valid ruby code but it's not meaningful, so we should fix it.
- Or, they made a mistake (for example, they had intended to write `Array.new(42) { 1 }`.

**Comment by ota42y** [comment]
I'm rebased to master

**Comment by bbatsov** [comment]
There can't be a mismatch between reported offenses and the behaviour of auto-correct. I'm a bit confused about your proposal for this.

**Comment by ota42y** [comment]
I just realized, I think there is a mismatch too.
So, I fixed offenses rule.

File: CHANGELOG.md [hunk 4]
```diff
@@ -19,6 +19,7 @@
 * [#4038](https://github.com/bbatsov/rubocop/pull/4038): Allow `default` key in the `Style/PercentLiteralDelimiters` cop config to set all preferred delimiters. ([@kddeisz][])
 * Add `IgnoreMacros` option to `Style/MethodCallWithArgsParentheses`. ([@drenmi][])
 * [#3937](https://github.com/bbatsov/rubocop/issues/3937): Add new `Rails/ActiveSupportAliases` cop. ([@tdeo][])
+* [#4049](https://github.com/bbatsov/rubocop/pull/4049): Bugfix for `Style/EmptyLiteral` cop. ([@ota42y][])
```
Comment (bbatsov) [inline CHANGELOG.md]: Why is this at two places in the changelog?

File: spec/rubocop/cop/style/empty_literal_spec.rb [hunk 5]
```diff
@@ -27,6 +27,24 @@
       new_source = autocorrect_source(cop, 'test = Array.new')
       expect(new_source).to eq('test = []')
     end
+
+    it 'does not registers an offence Array.new with block' do
+      source = 'test = Array.new { 1 }'
+      inspect_source(cop, source)
+      expect(cop.offenses.size).to eq(0)
```
Comment (bbatsov) [inline spec/rubocop/cop/style/empty_literal_spec.rb]: Use `be_empty` here.

File: lib/rubocop/cop/style/empty_literal.rb [hunk 6]
```diff
@@ -35,8 +37,12 @@ def on_send(node)
         end
 
         def autocorrect(node)
+          result = correction(node)
+
+          return unless result
```
Comment (bbatsov) [inline lib/rubocop/cop/style/empty_literal.rb]: I'm a bit confused here. How come a node reported as offence not be correctable?

**Comment by bbatsov** [comment]
Btw, `Hash.new` can also be passed a block, so I guess this should be addressed by this PR as well.

**Comment by ota42y** [comment]
I did fix!

File: spec/rubocop/cop/style/empty_literal_spec.rb [hunk 7]
```diff
@@ -31,19 +31,19 @@
...
       inspect_source(cop, source)
-      expect(cop.offenses.size).to eq(0)
+      expect(cop.offenses).to be_empty
     end
 
     it 'registers an offence when Array.new in block' do
       source = 'puts { Array.new }'
-      inspect_source(cop, source)
-      expect(cop.offenses.size).to eq(1)
+      new_source = autocorrect_source(cop, source)
```
Comment (bbatsov) [inline spec/rubocop/cop/style/empty_literal_spec.rb]: I think it was a mistake to update the code here. This test doesn't test auto-correct after all.

File: lib/rubocop/cop/style/empty_literal.rb [hunk 8]
```diff
@@ -87,12 +80,17 @@ def offence_array_node?(node)
           array_node(node) && !array_with_block(node.parent)
         end
 
+        def offence_hash_node?(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/empty_literal.rb]: Btw, we use the spelling `offense` everywhere.

File: spec/rubocop/cop/style/empty_literal_spec.rb (hunk 7 above)
Comment (ota42y) [inline spec/rubocop/cop/style/empty_literal_spec.rb]: Sorry, I can't understand... 

```
      source = 'puts { Array.new }' # source is non-auto-correct code
      new_source = autocorrect_source(cop, source) # do auto-corrected and save to new_source
      expect(new_source).to eq 'puts { [] }' # check new_source is correct ?
```

Comment (bbatsov) [inline spec/rubocop/cop/style/empty_literal_spec.rb]: This test is just supposed to check if there's offense in the code or not, it's not supposed to check if the auto-correct works.

Comment (ota42y) [inline spec/rubocop/cop/style/empty_literal_spec.rb]: Oh, I see! 
I'll fix test description.

**Pull request merged**